File: c++/longestSubstringWithoutRepeatingCharacters/longest_substring_without_repeating_characters.cpp
#include <map>
#include <algorithm>

#include "longest_substring_without_repeating_characters.h"

int lengthOfLongestSubstring(string str)
{
    map<char, int> occur;
    auto length = 0;
    auto longest = 0;
    auto startIndex = 0;
    for(auto index = 0; index < str.length(); index++)
    {
        auto c = str[index];
        auto iter = occur.find(c); 
        if(iter == occur.end())
        {
            occur[c] = index;
            length += 1;
        }
        else
        {
            auto lastIndex = iter->second;
            for(auto i = startIndex; i < lastIndex; i++)
            {
                auto deleteIter = occur.find(str[i]);
                occur.erase(deleteIter);
            }
            if(length > longest) longest = length;
            startIndex = lastIndex + 1;
            length = index - lastIndex;
            occur[c] = index;
        }
    }

    if(length > longest) longest = length;

    return longest;
}

File: elixir/longestSubstringWithoutRepeatingCharacters/lib/longestSubstringWithoutRepeatingCharacters.ex
defmodule LongestSubstringWithoutRepeatingCharacters do
    def lengthOfLongestSubstring(str) do
        {longest, length, _, _} = str |> String.to_char_list
        |> Enum.with_index
        |> Enum.reduce({0, 0, 0, %{}}, fn({c, index}, {longest, length, startIndex, occur}) ->
            lastIndex = Dict.get(occur, c)
            if lastIndex == nil do
                length = length + 1
            else
                occur = Enum.reduce(startIndex..lastIndex, occur, fn(i, acc) ->
                    Dict.delete(acc, String.at(str, i))
                end)
                if length > longest, do: longest = length
                startIndex = lastIndex + 1
                length = index - lastIndex
            end
            occur = Dict.put(occur, c, index)

            {longest, length, startIndex, occur}
        end)

        if length > longest, do: longest = length
        longest
    end
end
